js_modules/dagster-ui/packages/ui-core/src/asset-data/AssetLiveDataProvider.tsx

```diff
@@ -43,24 +43,52 @@ export function useAssetLiveData(
 export function useAssetsLiveData(
   assetKeys: AssetKeyInput[],
   thread: AssetLiveDataThreadID = 'default',
+  batchUpdatesInterval: number = 1000,
 ) {
   const [data, setData] = React.useState<Record<string, LiveDataForNode>>({});
+  const dataRef = React.useRef(data);
+  dataRef.current = data;
+
   const [isRefreshing, setIsRefreshing] = React.useState(false);
 
   const client = useApolloClient();
   const manager = AssetLiveDataThreadManager.getInstance(client);
 
   React.useEffect(() => {
-    const setDataSingle = (stringKey: string, assetData?: LiveDataForNode) => {
-      setData((data) => {
-        const copy = {...data};
-        if (!assetData) {
-          delete copy[stringKey];
-        } else {
+    let timeout: ReturnType<typeof setTimeout> | null = null;
+    let didUpdateOnce = false;
+    let didScheduleUpdateOnce = false;
+    let updates: {stringKey: string; assetData: LiveDataForNode | undefined}[] = [];
+
+    function processUpdates() {
+      const copy = {...dataRef.current};
+      updates.forEach(({stringKey, assetData}) => {
+        if (assetData) {
           copy[stringKey] = assetData;
+        } else {
+          delete copy[stringKey];
         }
-        return copy;
       });
+      updates = [];
+      setData(copy);
+    }
+
+    const setDataSingle = (stringKey: string, assetData?: LiveDataForNode) => {
+      /**
+       * Throttle updates to avoid triggering too many GCs and too many updates when fetching 1,000 assets,
+       */
+      updates.push({stringKey, assetData});
+      if (!didUpdateOnce) {
+        if (!didScheduleUpdateOnce) {
+          didScheduleUpdateOnce = true;
+          requestAnimationFrame(() => {
+            processUpdates();
+            didUpdateOnce = true;
+          });
+        }
+      } else if (!timeout) {
+        timeout = setTimeout(processUpdates, batchUpdatesInterval);
+      }
```

Comment:
This all looks good to me, but I wonder if we could break this functionality into a separate hook to keep this one from getting even more complex...  I think it'd be possible to wrap it all up into a `const [data, setData] = useStateWithThrottledSetter()` type of thing maybe?

Could also just wait until we need this somewhere else, I'm sure it was tough to get this working as-is